### ISSUE TITLE
MongoDB disconnect during backup in Meteor 1.4.1.1

### ISSUE THREAD

**Issue by mjmasn** [meteor/meteor#7783]
Since upgrading to Meteor 1.4.1.1 (I think this updated the mongo driver from 2.2.8 to 2.2.9) we've been having issues with our MongoDB replicaset on Compose. Each day when the backups run we're seeing a database disconnect in our apps. I've been in contact with Compose but they confirmed that the backups run on a separate hidden member so unless the driver is faulty, it shouldn't be possible for a backup to cause this kind of disconnect. I have confirmed that rolling back the code does resolve the issue.

Now in theory a brief disconnect doesn't matter to Meteor but it actually kills all our percolate:synced-cron jobs and they won't run until the app is restarted. (That's a separate bug I might try to have a look at next week.)

This may well be a bug in the latest mongo driver but I'm submitting this just in case anything else relating to mongo has been recently changed in Meteor that could be causing this issue.

**Comment by abernix** [comment]
Hopefully this is resolved by #7780 which updates the Mongo driver to the latest version and resolves what sounds like a very similar issue for another user.

If you have a way of reproducing this problem reliably, I encourage you to try [running meteor from a checkout](https://github.com/meteor/meteor#slow-start-for-developers) to see if the latest change resolves it for you!

**Comment by mjmasn** [comment]
I had already tried using the latest mongo driver 2.2.10 but also just tried running from checkout - still no joy. I've now updated Mongo from 2.6.9 to 3.0.11 and it seems to be resolved. I'll close this but worth knowing if anyone else with an older Meteor app stumbles on this one :)

**Comment by mjmasn** [comment]
Just to make this even more weird... 

The problem has returned this morning as the automated backup ran (but still works fine when I click the manual backup button on compose.io). On our production servers we're now seeing a different error - basically the previously uncaught Mongo error wrapped in a Meteor one, so it seems to recover ok (unless that's a weird coincidence and the timeout happened at a slightly different point in the code where the error could be caught). On my local machine (connected to another database in the same compose deployment) the old problem has resurfaced exactly as before, causing the cron jobs stopped working. (This is all on Meteor 1.4.1.1). 

Unfortunately the app I'm testing on a checkout of the devel branch had crashed with `Error: connect ENETUNREACH` just before the backup started so I couldn't see if the latest driver would have had any effect. Guess I'll find out tomorrow morning...

**Comment by Batistleman** [comment]
I think I'm seeing the same error, also one a day, luckily at night. (also connected to compose.io)

Kadira logs the following [server-crash]: 

    Error: write after end
        at writeAfterEnd (_stream_writable.js:166:12)
        at Socket.Writable.write (_stream_writable.js:211:5)
        at Socket.write (net.js:642:40)
        at [object Object].Connection.write (/opt/azumuta/app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:499:58)
        at /opt/azumuta/app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:193:16
        at Object.cb (/opt/azumuta/app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/auth/mongocr.js:100:11)
        at [object Object].<anonymous> (/opt/azumuta/app/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/pool.js:233:16)
        at [object Object].g (events.js:260:16)
        at emitTwo (events.js:87:13)
        at [object Object].emit (events.js:172:7)

And a different error (don't know if it is related) without a stack:

    message: Got exception while polling query: MongoError: pool destroyed 
    stack:

**Comment by tmeasday** [comment]
@mjmasn what are the error messages you see when it disconnects? Are you seeing the same thing as https://github.com/meteor/meteor/issues/7822?

**Comment by mjmasn** [comment]
@tmeasday no, it was just a `connection timed out` error at exactly 08:04 BST every morning (backups started at ~08:03). It hasn't happened again since 1 day after upgrading to Mongo 3.0.11 (as in the error appeared once, but only once, for each of our 5 apps with Mongo 3). I wonder if there was a temporary problem with Compose they haven't mentioned, it all happened over a ~1 week period...

This can probably be closed unless anybody can replicate it.

<details>
<summary>Click for stack traces.</summary>
This was the original error:
```
Exception in setTimeout callback

MongoError: connection 1 to x.y.z.mongolayer.com:9999 timed out
    at Object.Future.wait (/Users/mike/.meteor/packages/meteor-tool/.1.4.1_1.1d3gufh++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:449:15)
    at MongoConnection.<anonymous> (packages/meteor/helpers.js:119:1)
    at MongoConnection.(anonymous function) (packages/mongo/mongo_driver.js:771:49)
    at MongoConnection.mongoConnectionProto.(anonymous function) [as insert] (packages/meteorhacks_kadira/lib/hijack/db.js:62:1)
    at OptimizedApply (packages/meteorhacks_kadira/lib/utils.js:31:1)
    at Object.ret.(anonymous function) [as insert] (packages/meteorhacks_kadira/lib/hijack/db.js:20:1)
    at [object Object].insert (packages/mongo/collection.js:528:37)
    at packages/percolate_synced-cron/synced-cron-server.js:203:1
    at scheduleTimeout (packages/percolate_synced-cron/synced-cron-server.js:266:1)
    at packages/percolate_synced-cron/synced-cron-server.js:314:1
    - - - - -
    at Function.MongoError.create (/Users/mike/.meteor/packages/npm-mongo/.1.5.49.1ltgr8n++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/error.js:29:11)
    at Socket.<anonymous> (/Users/mike/.meteor/packages/npm-mongo/.1.5.49.1ltgr8n++os+web.browser+web.cordova/npm/node_modules/mongodb-core/lib/connection/connection.js:173:20)
    at Socket.g (events.js:260:16)
    at emitNone (events.js:67:13)
    at Socket.emit (events.js:166:7)
    at Socket._onTimeout (net.js:332:8)
    at _runOnTimeout (timers.js:524:11)
    at _makeTimerTimeout (timers.js:515:3)
    at Timer.unrefTimeout (timers.js:584:5)
```

After upgrading to Mongo 3, Meteor seemed to be catching the error in a slightly different way (may just be the difference between running `meteor` and running a bundled app on the server though. If it's important I can dig out the archived logs, we've only got a week of searchable logs on papertrail and we updated to Mongo 3 exactly a week ago!):
```
Exception while polling query {"collectionName":"users","selector":{"_id":"xxx"},"options":{"reactive":true,"limit":1}}: MongoError: connection 227 to x.y.z.mongolayer.com:9999 timed out 

    at Object.Future.wait (/x/x/x/bundle/programs/server/node_modules/fibers/future.js:449:15)
    at SynchronousCursor._nextObject (packages/mongo/mongo_driver.js:1021:47)
    at SynchronousCursor.forEach (packages/mongo/mongo_driver.js:1055:22)
    at SynchronousCursor.getRawObjects (packages/mongo/mongo_driver.js:1104:12)
    at PollingObserveDriver._pollMongo (packages/mongo/polling_observe_driver.js:152:48)
    at PollingObserveDriver.proto._pollMongo (packages/meteorhacks_kadira/lib/hijack/wrap_observers.js:100:1)
    at Object.task (packages/mongo/polling_observe_driver.js:90:12)
    at [object Object]._.extend._run (packages/meteor/fiber_helpers.js:147:1)
    at packages/meteor/fiber_helpers.js:125:1
    at Function.MongoError.create (/x/x/x/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/error.js:29:11)
    at Socket.<anonymous> (/x/x/x/bundle/programs/server/npm/node_modules/meteor/npm-mongo/node_modules/mongodb-core/lib/connection/connection.js:173:20)
    at Socket.g (events.js:260:16)
    at emitNone (events.js:67:13)
    at Socket.emit (events.js:166:7)
    at Socket._onTimeout (net.js:332:8)
    at _runOnTimeout (timers.js:524:11)
    at _makeTimerTimeout (timers.js:515:3)
    at Timer.unrefTimeout (timers.js:584:5)
```

</details>

**Comment by tmeasday** [comment]
Thanks @mjmasn! I'll close it as I don't think we've seen any other reports.

@Batistleman your issue sounds like a different problem again..?

**Comment by abernix** [comment]
@Batistleman Just in case this helps:

***

A new version of the `npm-mongo` package (now 1.5.50) has been published.  This Meteor package includes the latest version of the Mongo Node driver, 2.2.10 ([View their bug-fixes here](https://github.com/mongodb/node-mongodb-native/blob/2.2/HISTORY.md#2210-2016-09-15)).

Maybe worth a try to see if its related to something in the underlying driver.  You can update to this new package by running `meteor update npm-mongo` in your Meteor project:

```
user@host:~/project_dir $ meteor update npm-mongo

Changes to your project's package version selections from updating package versions:

npm-mongo  upgraded from 1.5.49 to 1.5.50
```

You should see the change to 1.5.50.

*** 

If you continue to have problems, please open a new issue since your issue appears to be different. :smile:

**Comment by Batistleman** [comment]
Thanks for the tip! As @mjmasn says... I haven't seen the error in a while now. I also think it might have been a temp compose problem.
